tests/js/tags/admin.js

```diff
@@ -0,0 +1,498 @@
+    expect(result).toBe(113);
+    expect(table.instance().visible.length).toBe(0);
+    expect(table.instance().setState.mock.calls).toEqual([[{"checked": new Set()}]]);
+});
+
+
+test('TagResourceTable handleTableResize', () => {
+    const table = shallow(<TagResourceTable />);
+    table.instance().visible = [23, 43];
+    table.setState({checked: new Set([23, 43, 73])});
+    table.instance().setState = jest.fn(() => 113);
+    let result = table.instance().handleTableResize(5);
+    expect(result).toBe(113);
+    // visible is filled to pageSize
+    expect(table.instance().visible).toEqual(
+        [23, 43, undefined, undefined, undefined]);
+    expect(table.instance().setState.mock.calls).toEqual([[{"checked": new Set([23, 43])}]]);
+});
+
+
+test('TagResourceTable pruned', async () => {
+    const table = shallow(<TagResourceTable />);
+    let result = table.instance().pruned;
+    expect(result).toEqual(new Set([]));
+    table.instance().visible = [1, 3, 5];
+    table.setState({checked: new Set([0, 1, 2, 3])});
+
+    result = table.instance().pruned;
+    expect(result).toEqual(new Set([1, 3]));
+});
+
+
+test('TagResourceTable handleSubmit', async () => {
+    const handleSubmit = jest.fn(async () => 23);
+    const table = shallow(<TagResourceTable handleSubmit={handleSubmit} />);
+    const evt = {preventDefault: jest.fn()};
+    await table.instance().handleSubmit(evt);
+    expect(handleSubmit.mock.calls).toEqual([[[]]]);
+    expect(evt.preventDefault.mock.calls).toEqual([[]]);
+    expect(table.state().checked).toEqual(new Set());
+
+    table.instance().state.checked = new Set([1, 2, 3]);
+    await table.instance().handleSubmit(evt);
+    expect(handleSubmit.mock.calls[1]).toEqual([[1, 2, 3]]);
+    expect(evt.preventDefault.mock.calls[1]).toEqual([]);
+    expect(table.state().checked).toEqual(new Set());
+});
+
+
+test('TagResourceTable handleCheckboxClick', async () => {
+    const table = shallow(<TagResourceTable />);
+
+    const evt = {preventDefault: jest.fn(), target: {checked: false, name: 23}};
+    table.instance().handleCheckboxClick(evt);
+    const checked = new Set();
+    expect(table.state()).toEqual({checked});
+
+    evt.target.checked = true;
+    table.instance().handleCheckboxClick(evt);
+    checked.add(23);
+    expect(table.state()).toEqual({checked});
+
+    table.instance().handleCheckboxClick(evt);
+    expect(table.state()).toEqual({checked});
+
+    evt.target.checked = false;
+    table.instance().handleCheckboxClick(evt);
+    checked.delete(23);
+    expect(table.state()).toEqual({checked});
+});
+
+
+test('TagResourcesLoader constructor', () => {
+    document.querySelectorAll = jest.fn(() => new Set([1, 2, 3]));
+    const loader = new TagResourcesLoader();
+    const nodes = loader.tagResourceNodes;
+    expect(nodes).toEqual([1, 2, 3]);
+    expect(document.querySelectorAll.mock.calls).toEqual(
+        [["div.js-tag-resources"]]);
+})
```

Comment:
I've got good comments on the patterns here below on the test on 477ff. Sorry for the reverse-polish notation